### PR TITLE
Document "wrap" validators

### PR DESCRIPTION
**Pull request by tpdorsey**
Add documentation for before, after, and wrap validators

## Related issue number

Closes PYD-98

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: docs/usage/validators.md [hunk 1]
```diff
@@ -368,3 +368,52 @@ print(DemoDataclass())
...
+
+You can also use the  `mode` argument for `@field_validator` and `@model_validator` to specify when validation occurs.
+
+
+```py
+from typing import Annotated
+
+from pydantic import AfterValidator, BaseModel, ValidationError
+
+MyInt = Annotated[int, AfterValidator(lambda v: v + 1)]
```
Comment (adriangb) [inline docs/usage/validators.md:392]: I think an example of `WrapValidator` would also be nice.

**Review by adriangb — APPROVED** [review]
I've found an area of confusion is the relationship between the order in `Annotated` or in the class definition (in this case inheritance is also relevant) and how the validators are run. E.g. how `Annotated[..., AfterValidator(func1), BeforeValidator(func2)]` `func2` will be called before `func1`.

Also worth explaining that you can use multiple before/after/wrap validators but only 1 plain validator.

File: docs/usage/validators.md (hunk 1 above)
Comment (tpdorsey) [reply docs/usage/validators.md]: @adriangb good callouts. I added these examples, and also consolidated my new section with "Pre and per-item validators" to reduce redundancy.

File: docs/usage/validators.md [hunk 2]
```diff
@@ -159,12 +175,18 @@ except ValidationError as e:
     """
 ```
 
+Note that the type of validator dictates the order of validation, not the order in which validators are applied in an `Annotated` type or class definition. In the following example, `func2` will be called before `func1`.
```
Comment (adriangb) [inline docs/usage/validators.md:178]: Need to clarify that within a given type they go from right to left and back. That is it goes from right to left running all before validators (or calling into wrap validators) and then left to right back out calling all after validators. Doesn’t need to be in this paragraph, it’s just that as it stands right now this says “the order in Annotated doesn’t matter” but it does.

File: docs/usage/validators.md [hunk 3]
```diff
@@ -195,7 +217,43 @@ print(DemoModel(int_list=[3, 2, 1], name_list=['adrian g', 'David']))
 #> int_list=[1, 2, 3] name_list=['Adrian G', 'David']
 ```
 
-## Validate Always
+### Wrap validators
+
+Wrap validators are useful for cases where you want to validate a value, but also want to return a default value if validation fails.
+
+```py
```
Comment (adriangb) [inline docs/usage/validators.md:224]: Did this come from a test? If so we can delete the test

**Review by adriangb — APPROVED** [review]
Just some small notes. Please merge whenever you feel ready.

File: docs/usage/validators.md (hunk 3 above)
Comment (tpdorsey) [reply docs/usage/validators.md:224]: No this is adapted from the docstring

**Pull request merged**